File: data_analysis/count_vowels.py
"""Count vowels in strings (Python).

Includes an `extended` option to normalize Unicode and count accented vowels
as base vowels.
"""
from typing import Optional
import unicodedata


def count_vowels(s: str, extended: bool = True) -> int:
    """Count vowels in a string.

    Args:
        s: input string
        extended: if True, normalize Unicode (NFD) and treat accented vowels as
                  their base (e.g. é -> e). If False, count ASCII vowels only.

    Returns:
        Number of vowels (int)
    """
    if not isinstance(s, str):
        raise TypeError("s must be a string")

    vowels = set("aeiou")
    if extended:
        normalized = unicodedata.normalize("NFD", s)
        count = 0
        for ch in normalized:
            if unicodedata.category(ch) == "Mn":
                continue
            if ch.lower() in vowels:
                count += 1
        return count

    return sum(1 for ch in s.lower() if ch in vowels)


if __name__ == "__main__":
    examples = [
        ("Hello, World!", False),
        ("Résumé", True),
        ("ÁÉÍÓÚáéíóú", True),
        ("", True),
    ]
    for text, ext in examples:
        print(repr(text), "extended=", ext, "->", count_vowels(text, extended=ext))

File: data_analysis/count_vowels.R
# count_vowels.R
# Count vowels in a string with optional Unicode normalization

#' Count vowels in a string
#'
#' @param s character(1) input string
#' @param extended logical(1) If TRUE and stringi is available, normalize and
#'                 strip diacritics so accented vowels are counted as their base.
#' @return integer count of vowels (a/e/i/o/u)
count_vowels <- function(s, extended = TRUE) {
  if (!is.character(s) || length(s) != 1) stop("s must be a single string")

  if (isTRUE(extended) && requireNamespace("stringi", quietly = TRUE)) {
    s2 <- stringi::stri_trans_nfkd(s)
    s2 <- gsub("\\p{M}", "", s2, perl = TRUE)
  } else {
    if (isTRUE(extended)) {
      warning("stringi not available; falling back to ASCII-only matching")
    }
    s2 <- s
  }

  matches <- gregexpr("[aeiouAEIOU]", s2, perl = TRUE)[[1]]
  if (length(matches) == 1 && matches[1] == -1) return(0L)
  length(matches)
}

# Examples
# print(count_vowels("Hello, World!"))  # -> 3
# print(count_vowels("Résumé", extended = TRUE))  # -> 3 (if stringi available)
